### ISSUE TITLE
How to compile for iOS?

### ISSUE THREAD

**Issue by Joride** [libexif/libexif#80]
I am interested in including libexif in an iOS application. Either include the source directly, or build a static lib.

I cloned the repository, and tried to follow the instructions from readme
Unfortunately, the instructions do not work on my machine (M1 MacBook Air, running macOS Monterey (12.0.1 (21A559)).
As per the readme:
_When building from source out of git, something like the following will be
necessary:

  autoreconf -i
  ./configure
  make_

running  `autoreconf -I` or `./ autoreconf -I` give the following output: _zsh: command not found: autoreconf_

running `./configure` also errors with `zsh: no such file or directory: ./configure`

Including the source files from the _libexif_ directory in an Xcode project won't work, as  `config.h` will be missing.

Could you please advice?

**Comment by dfandrich** [comment]
If you build from a release tarball, the autoreconf step is already done for
you so you can skip to the ./configure step. Otherwise, you'll need to install
the GNU autotools package to get autoreconf.

**Comment by ndim** [comment]
I am wondering... isn't there an exif parsing library already in a more or less native to iOS language and form?

**Comment by Joride** [comment]
@ndim yes, the Image I/O framework on iOS can read exif and maker notes for a number of raw files (I tested CR2 and NEF). But the Image I/O framework cannot write those types of raw files (`'error 119: unsupported file format 'com.canon.cr2-raw-image''`), the list of supported  types for writing is: public.jpeg, public.png, com.compuserve.gif, public.tiff, public.jpeg-2000, com.apple.atx, org.khronos.ktx, org.khronos.astc, com.microsoft.dds, public.heic, public.heics, com.microsoft.ico, com.microsoft.bmp, com.apple.icns, com.adobe.photoshop-image, com.adobe.pdf, com.truevision.tga-image, com.ilm.openexr-image, public.pbm, public.pvr.

On macOS, I could use exiftool (https://exiftool.org), as from a Mac app I can spawn a separate process that would run the perl script that is exiftool, using the perl installation that comes with macOS. I looked into embedding Perl into an app itself for iOS (as iOS has no perl built in), but alas, I ran into issues trying to compile Perl into a static library.

Then I found this repository, but since I lack a good understanding of the configure/make/install commands, I have trouble getting it to a state where I can use it from within an app.

**Comment by Joride** [comment]
@dfandrich Thanks, that sounds helpful, I will try starting from the tarball. Could you point me to a place where I can learn about the options for configure/make to create a library that can be used on iOS (arm64)?

**Comment by ndim** [comment]
Have you read the `README` and `INSTALL` files?

  * If you start from a git clone, you need the autoconf/automake tools installed to build the buildsystem (including the `configure` sh script) first. If you start from a release tarball, this step has already been done.
  * You run `./configure --help` for a list of configuration options, and then run `configure CC=aarch64-whatever-ios-cc --prefix=/path/to/wherever` to generate a set of `Makefile`s configured for your specific build. This step will also generate a `config.h` file containing definitions specific to that build.
  * You run `make` to build the library.
  * You may run `make install to install the built library into the `--prefix` you have given.

All of this will require the appropriate tools to be installed on your build system (`make`, compilers, etc.) and you calling `configure` and `make` with the appropriate tools being available and environment variables like PATH being set properly.

I am not aware of anyone here being familiar with building for iOS. The most Appley thing I have seen here is building on OSX for OSX.

**Comment by Joride** [comment]
@ndim Thank you for your quick response. Yes I read the readme, but not the install file. That one sure contains a lot of info I was looking for, not sure why I missed it.
Your instructions are exactly what I was looking for, I ran `./configure build=arm64-apple-ios15.0 --prefix=/Users/Joride/libexif` and I got a `.a` file for which `lipo -info` gives me `architecture: arm64` and the familiar 'include' folder. So this looks great! Next step is to put it in an app and see if there are any iOS portability issues. If so, I could propose a fix and create a PR for it.
Thanks again for your help, much appreciated!

**Comment by Joride** [comment]
I tried various incantations of ./configure, but I can't seem to get a build that is recognized as being built for iOS-arm64. When trying to compile in the the static lib, I get an error `Building for iOS, but the linked library 'libexif.a' was built for macOS`. Unfortunately, this is an M1 Mac I am using, which is also arm64. So I need to run the same command on an old Intel Mac, see if that would give me x86_64. If that would be the case, somehow my ./configure commands do not result in a built for the right arch.

How realistic is it to avoid the use of configure/make entirely, and invoke the compiler commands myself? (which means using Xcode and setting the compiler settings in the build settings there)

**Comment by ndim** [comment]
> [...] I ran `./configure build=arm64-apple-ios15.0 --prefix=/Users/Joride/libexif` and I got a `.a` file for which `lipo -info` gives me `architecture: arm64` and the familiar 'include' folder. [...]

> I tried various incantations of ./configure, but I can't seem to get a build that is recognized as being built for iOS-arm64. When trying to compile in the the static lib, I get an error `Building for iOS, but the linked library 'libexif.a' was built for macOS`.

Well, I would be very surprised if the configure line you were quoting above would result in a working cross-compile. First, the argument name is `--build=`, not `build=`. Second, quoting `configure --help`:
```
System types:
  --build=BUILD     configure for building on BUILD [guessed]
  --host=HOST       cross-compile to build programs to run on HOST [BUILD]
```
you were *not* setting the system you want the built library to run on. The name for that would be `--host`.

Given that you appear to be confusing both the very different strings `--build=` and `build=`, and also the BUILD and HOST concepts, I would recommend you double and triple check that `arm64-apple-ios15.0` is actually the proper string to use here.

For gcc, cross compilers are usually executables called something like `aarch64-linux-gnu-gcc` when they generate code to run on the `aarch64-linux-gnu` system. clang might be a little different there, and may be using a system of `clang --some-target-option=aarch64-linux-gnu` instead. That might mean that you need to give `CC="clang --some-target-option=aarch64-linux-gnu"` as an argument to `configure`.

> Unfortunately, this is an M1 Mac I am using, which is also arm64. So I need to run the same command on an old Intel Mac, see if that would give me x86_64. If that would be the case, somehow my ./configure commands do not result in a built for the right arch.

It is not just a question of the processor architecture, but also of the system libraries and APIs. I would presume there are significant differences here as well between iOS and OSX. So you need to refine the command examining the resulting library binary to distinguish between iOS and OSX builds. I am certain there is such a command.

> How realistic is it to avoid the use of configure/make entirely, and invoke the compiler commands myself? (which means using Xcode and setting the compiler settings in the build settings there)

No idea. At the very least, you would need to manually do what the `configure` script does to generate a `config.h` file from `config.h.in` which does sound like a bit of research work I personally would shy away from.

The GNU Makefile standard and an Autoconf-generated configure script are standards which have been working for a lot of software to build on lots of systems over more than the last two decades. E.g. the GNU compiler collection (gcc) uses that system, and you can definitively build gcc on a lot of systems to run on a lot of systems and to build programs which run on even more systems. If you want to do your own thing, you are welcome to do your own thing, but as far as I know the core libexif team, I would be surprised if anyone would be in a position to offer you much help on using XCode.

**Comment by Joride** [comment]
@ndim Thank you for your patience, and very clear explanation. I have been able to build a static library for iOS (arm64). Putting them here for future reference and to benefit others.

I create a `configureiOS.sh` file, made it executable Added this to the directory containing the contents of the tarball. This `configureiOS.sh` contains:

`#!/bin/sh -xe

libDirName="exiflib"

ARCH=arm64-apple-darwin10
export ARCH

clang="/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang"
sdk="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.0.sdk"

IPHONEOS_DEPLOYMENT_TARGET=${IPHONEOS_DEPLOYMENT_TARGET:-15.0}
export IPHONEOS_DEPLOYMENT_TARGET

CC="$clang"
export CC

CFLAGS="-target $ARCH --sysroot=$sdk $optflags -mios-version-min=$IPHONEOS_DEPLOYMENT_TARGET"
export CFLAGS

host=aarch64-apple-darwin10

mkdir /Users/$(whoami)/$libDirName

./configure --host="$host" --prefix=/Users/$(whoami)/$libDirName --disable-shared --disable-dependency-tracking
`

Now I can run `./configureiOS.sh`, followed by
`make install`
And I get `.a` file and the expected 'include' folder.

Running `lipo -info libexif.a` outputs: `Non-fat file: libexif.a is architecture: arm64`
Running `otool -l libexif.a | grep platform -m 1` outputs `platform 2`, indicating linkage against the iOS SDK.

This will get me started on an actual app. 
For those of us looking to be able to work on simulator (for either Intel and/or Apple silicon), look into the architectures that are required for those, update the SDK paths to the correct ones, and use `lipo` to create 'fat binaries'. 
I could not find official docs listing the architecture name (in the script above 'arch64-apple-darwin10` and `arm64-apple-darwin10`), but these can be found (tediously) by creating a new Xcode project for static library, and building it for the various architecture. Then inspect the details of the build commands in the Report Navigator of Xcode.
Some googling showed me these (in braces my assumptions on what devices can be targeted with those)

`i386-apple-darwin11` = iPhone Simulator on Intel
`armv6-apple-darwin10` (ancient iPhones)
`armv7-apple-darwin10` = (very old iPhones)
`armv7s-apple-darwin10` = (old iPhones)
`arm64-apple-darwin10` = (iPhones, starting from iPhone 5S)

**Comment by ndim** [comment]
If you don't mind, I want to keep this open for a bit, do a few experiments based on that script on a friend's MacOS machine, and then document the results in `README` or another file as part of the libexif documentation.

**Comment by Joride** [comment]
@ndim yeah, sounds good! I might try and work out the details on supporting simulator (on M1 and Intel macs) this week. I will update here with the results.

**Comment by Joride** [comment]
I managed to build for arm64 devices (basically all iOS devices released since late 2013 ), and for both Intel and Apple Silicon simulator.

What follows is basically a slight adaptation of the script posted [here](https://stackoverflow.com/a/26812514/2358592), it builds 1 archive for the devices (as all devices have the same architecture. The simulator is considered a different platform and it requires either `x86_64` for Intel macs, or `arm64` for Apple Silicon (currently that means M1) macs. These two architecture can be combined into one archive using `lipo`.

The script below is a prerequisite for setting up an application in Xcode to use _and debug_ this library, as described [here](https://github.com/libexif/libexif/issues/82#issuecomment-984996628).

```
#!/bin/sh -xe

# adapted from the script found here:
# https://stackoverflow.com/a/26812514/2358592

OUTPUT_DIR=$PWD/..

# `-Os`            : optimize for size
# `-fembed-bitcode`: embed llvm bitcode
# `-g3`            : produce debugging information
OPT_FLAGS="-Os -fembed-bitcode -g3"

# on M1 Macbook Air there are 8 cores, so that seems an appropriate number
MAKE_JOBS=8

dobuild()
{
    export CC="$(xcrun -find -sdk ${SDK} cc)"
    export CXX="$(xcrun -find -sdk ${SDK} cxx)"
    export CPP="$(xcrun -find -sdk ${SDK} cpp)"
    export CFLAGS="${HOST_FLAGS} ${OPT_FLAGS}"
    export CXXFLAGS="${HOST_FLAGS} ${OPT_FLAGS}"
    export LDFLAGS="${HOST_FLAGS}"

    ./configure --host=${CHOST} --prefix=${PREFIX} --enable-static --disable-shared

    make clean
    make -j${MAKE_JOBS}
    make install
}

# all iOS devices released after late 2013
SDK="iphoneos"
ARCH_FLAGS="-arch arm64"
HOST_FLAGS="${ARCH_FLAGS} -miphoneos-version-min=15.0 -isysroot $(xcrun -sdk ${SDK} --show-sdk-path)"
CHOST="arm-apple-darwin"
PREFIX="$OUTPUT_DIR/iphoneos-arm64"
dobuild

# Simulator on Intel macs
SDK="iphonesimulator"
ARCH_FLAGS="-arch x86_64"
HOST_FLAGS="${ARCH_FLAGS} -mios-simulator-version-min=15.0 -isysroot $(xcrun -sdk ${SDK} --show-sdk-path)"
CHOST="x86_64-apple-darwin"
PREFIX="$OUTPUT_DIR/iphonesimulator-x86_64"
dobuild

# Simulator on Apple Silicon (e.g. M1) macs
SDK="iphonesimulator"
ARCH_FLAGS="-arch arm64"
HOST_FLAGS="${ARCH_FLAGS} -mios-simulator-version-min=15.0 -isysroot $(xcrun -sdk ${SDK} --show-sdk-path)"
CHOST="arm-apple-darwin"
PREFIX="$OUTPUT_DIR/iphonesimulator-arm64"
dobuild

# create a universal('fat') binary for the two architecture for the iphonesimulator platform
mkdir -p "$OUTPUT_DIR/iphonesimulator-universal/include"
mkdir -p "$OUTPUT_DIR/iphonesimulator-universal/lib"
mkdir -p "$OUTPUT_DIR/iphonesimulator-universal/share"
cp -R "$OUTPUT_DIR/iphonesimulator-arm64/include/" "$OUTPUT_DIR/iphonesimulator-universal/include/"
cp -R "$OUTPUT_DIR/iphonesimulator-arm64/share/" "$OUTPUT_DIR/iphonesimulator-universal/share/"
lipo -create "$OUTPUT_DIR/iphonesimulator-arm64/lib/libexif.a" "$OUTPUT_DIR/iphonesimulator-x86_64/lib/libexif.a" -output "$OUTPUT_DIR/iphonesimulator-universal/lib/libexif.a"

# delete the non-universal folders
rm -rf "$OUTPUT_DIR/iphonesimulator-arm64"
rm -rf "$OUTPUT_DIR/iphonesimulator-x86_64"

```
